[`airflow`] Passing positional argument into `airflow.lineage.hook.HookLineageCollector.create_asset` is not allowed (`AIR303`)

sjyangkevin opened the pull request:
<!--
Thank you for contributing to Ruff/ty! To help us out with reviewing, please consider the following:

- Does this pull request include a summary of the change? (See below.)
- Does this pull request include a descriptive title? (Please prefix with `[ty]` for ty pull
  requests.)
- Does this pull request include references to any relevant issues?
-->

## Summary

This is a follow up PR to https://github.com/astral-sh/ruff/pull/21096

The new code AIR303 is added for checking function signature change in Airflow 3.0. The new rule added to AIR303 will check if positional argument is passed into `airflow.lineage.hook.HookLineageCollector.create_asset`. Since this method is updated to accept only keywords argument, passing positional argument into it is not allowed, and will raise an error. The test is done by checking whether positional argument with 0 index can be found.

## Test Plan

A new test file is added to the fixtures for the code AIR303. Snapshot test is updated accordingly.

<img width="1444" height="513" alt="Screenshot from 2025-12-17 20-54-48" src="https://github.com/user-attachments/assets/bc235195-e986-4743-9bf7-bba65805fb87" />

<img width="981" height="433" alt="Screenshot from 2025-12-17 21-34-29" src="https://github.com/user-attachments/assets/492db71f-58f2-40ba-ad2f-f74852fa5a6b" />

Lee-W commented on crates/ruff_linter/src/checkers/ast/analyze/expression.rs:
```diff
@@ -1278,6 +1271,9 @@ pub(crate) fn expression(expr: &Expr, checker: &Checker) {
                 airflow::rules::airflow_3_0_suggested_update_expr(checker, expr);
             }
+            if checker.is_rule_enabled(Rule::Airflow3FunctionSignatureChange) {
+                airflow::rules::airflow_3_keyword_args_only_function(checker, expr);
```
We probably can make a parent function like `airflow_3_function_signature_change_expr` and call `airflow_3_keyword_args_only_function` inside. IIRC, there're come other cases as well

ntBre commented on crates/ruff_linter/src/rules/airflow/rules/function_signature_change_in_3.rs:
```diff
@@ -0,0 +1,150 @@
+use ruff_text_size::Ranged;
+
+#[derive(ViolationMetadata)]
+#[violation_metadata(stable_since = "0.13.0")]
```
Good catch. I think 0.14.11 would be a good start, assuming this might land next week. We can always bump the version again just before landing, if needed.

sjyangkevin replied:
Thanks for catching it and thank @ntBre for the feedback. I think we could set it to 0.14.11, and bump the version if needed.

sjyangkevin commented:
Hi @Lee-W , @ntBre , I've made the following updates to the PR. The updates might be more than what we've discussed, but please let me know if you have further feedback. Thanks for your time to review.

### Updates
* created a parent function `airflow_3_function_signature_change_expr` to call `airflow_3_keyword_args_only_function`
* updated the `violation_metadata` to 0.14.11
* updated the struct document to be more generic for function signature change
* updated some namings and description for functions
* removed the check for `create_dataset`, which is checked by the `AIR301` for deprecation, implement only function signature change check for `create_asset` in `AIR303` (crates/ruff_linter/src/rules/airflow/rules/function_signature_change_in_3.rs **L107**)
* fixed the document generation issue, CI error for document generation should be fixed

> I'm not sure whether we want to make this a replacement. Or should we just make it a `Message`? Or this probably should be something like `make pos arg passed through keyword`, but we'll need to store the order of that function as well. it's not super hard but not sure whether fixing that worth the effort

I think again based on the following trade-off, I feel make it a `Message` probably a better option. @Lee-W let me know if you have different thoughts, or you think there is a better way to implement it.

Option 1: make it a message
* Implementation is straight-forward, and less maintenance efforts
* User experience might not be great, as they need to manually fix this, even though the fix is straight-forward
* Since it doesn't suggest any fix, no edge case to handle when suggesting/executing the fix

Option 2: make a "full replacement"
* In this case, we will need to store the order of the arguments for a function. Then every update (e.g., `create_asset(uri, extra)` to `create_asset(uri, metadata)`) would require an update to the rule. I am wondering if it could introduce long-term maintenance effort to ensure the rule work properly
* Wondering if the handling of starred args would be complicated (e.g., `collector.create_asset(*some_list, **some_dict)`). This might be an edge case. In `node.rs`, seems like there is a function can resolve the order for this case, so it could possibly be handled as well.

Lee-W commented on crates/ruff_linter/src/rules/airflow/rules/function_signature_change_in_3.rs:
```diff
@@ -0,0 +1,141 @@
+            function_def,
+            replacement,
+        } = self;
+        match replacement {
```
Do you thing we want a new enum or keep using `Replacement`? I feel like it's less likely to reuse things like `SourceModuleMoved`

Lee-W commented on crates/ruff_linter/src/rules/airflow/rules/function_signature_change_in_3.rs:
```diff
@@ -0,0 +1,141 @@
+        Some(Fix::safe_edit(Edit::range_replacement(
+            name.to_string(),
+            attr.range(),
+        )))
```
According to the description and the test output, we're not actually fixing anything. What is this for?

sjyangkevin replied:
Right. Since I am referencing the AIR301 implementation, and that contains the fix. I should actually remove it

sjyangkevin commented on crates/ruff_linter/src/rules/airflow/rules/function_signature_change_in_3.rs:
```diff
@@ -0,0 +1,117 @@
+    fn fix_title(&self) -> Option<String> {
+        let Airflow3FunctionSignatureChange { change_type, .. } = self;
+        match change_type {
+            FunctionSignatureChangeType::KeywordOnly { message } => Some(message.to_string()),
```
I am wondering whether we should put it onto the heap using `to_string()`.

The `message` is a fixed size string as defined on L98.

sjyangkevin replied:
I thinks as the definition `fn fix_title(&self) -> Option<String>`, we need to use `to_string()` to create a `String`.

ntBre commented on crates/ruff_linter/src/rules/airflow/rules/function_signature_change_in_3.rs:
```diff
@@ -0,0 +1,117 @@
+}
+
+/// AIR303
+pub(crate) fn airflow_3_function_signature_change_expr(checker: &Checker, expr: &Expr) {
```
nit: Do we need the `_expr` suffix here? I'd usually only expect that if the rule were also applied to statements and there were a `_stmt` variant.

ntBre commented on crates/ruff_linter/src/rules/airflow/rules/function_signature_change_in_3.rs:
```diff
@@ -0,0 +1,117 @@
+/// ## Why is this bad?
+/// Airflow 3.0 might introduce changes to function signatures. Code that
+/// worked in Airflow 2.x may raise a runtime error if not updated in Airflow
+/// 3.0.
```
What do you think about something like this?

```suggestion
/// Airflow 3.0 introduces changes to function signatures. Code that
/// worked in Airflow 2.x will raise a runtime error if not updated to work with Airflow
/// 3.0.
```

It seems like some function signatures have changed since y'all are adding this rule :)

ntBre reviewed (commented):
Thank you! This looks great to me overall, I just had a few small suggestions.

sjyangkevin commented:
Thanks @ntBre, I've made the following changes based on the feedback.

1. Renamed the test file from `AIR303_args.py` to `AIR303.py`
2. Renamed `airflow_3_function_signature_change_expr` to `airflow_3_function_signature_change`
> nit: I think I would probably just inline these functions:
3. Refactored the implementation to reduce the hierarchy of the function calls
4. Emitted the diagnostic directly to avoid the trail of `else { return; }` arms
5. Moved the enum `FunctionSignatureChangeType` into `function_signature_change_in_3.rs`, as it is less likely to be an enum shared by other modules/implementations. In case in future we need to tailor the message for different types of function signature changes, probably could still keep it.
6. Renamed `qualname` to `qualified_name` to avoid abbreviating names
7. Renamed `Airflow3FunctionSignatureChange` to `Airflow3IncompatibleFunctionSignature` to better align with naming convention
8. Updated the struct document for `Airflow3IncompatibleFunctionSignature` according to the feedback
9. Enhanced the check logic (use this condition `arguments.find_positional(0).is_some() || arguments.args.iter().any(Expr::is_starred_expr);`) to handle starred expressions (i.e., `*args` and `**kwargs`) ; the updated test cases and results are shown in `AIR303.py` and `ruff_linter__rules__airflow__tests__AIR303_AIR303.py.snap`.
10. Refactored the code to use `if let` instead of `match` as we only have a single pattern matching here. Using `match` raise a clippy warnings. We can refactor this to use `match` when more cases coming in the future.

```rust
if let ["airflow", "lineage", "hook", "HookLineageCollector"] = qualified_name.segments() {
    if attr.as_str() == "create_asset" && has_positional_args {
        ...
```
11. Update the `message` to show better description in the document.
<img width="1036" height="122" alt="Screenshot from 2025-12-27 00-51-14" src="https://github.com/user-attachments/assets/da4339b5-b764-45a3-8286-961231732d63" />

@Lee-W , would you mind have a look into these changes whenever you have time, and let me know if you have further feedback?

Thanks again for both your time in reviewing this!

ntBre commented on crates/ruff_linter/src/rules/airflow/rules/function_signature_change_in_3.rs:
````diff
@@ -0,0 +1,125 @@
+/// collector.create_asset(uri="s3://bucket/key")
+/// ```
+#[derive(ViolationMetadata)]
+#[violation_metadata(stable_since = "0.14.11")]
````
Rules have to start out in preview.


```suggestion
#[violation_metadata(preview_since = "0.14.11")]
```

ntBre commented on crates/ruff_linter/src/rules/airflow/rules/function_signature_change_in_3.rs:
```diff
@@ -0,0 +1,125 @@
+        return;
+    };
+
+    let qualified_name = if let Expr::Call(call) = value.as_ref() {
```
This looks good to me! Checking that `value` is a `Call` should be quite cheap, and I don't mind having the special case first.

sjyangkevin commented:
Thanks @ntBre , I've made the following updates based on your feedback.

1. Put the enum `FunctionSignatureChangeType` to the bottom of the file
2. Updated `stable_since` to `preview_since`
3. Renamed `airflow_3_function_signature_change` to `airflow_3_incompatible_function_signature` to match the new `Violation` name
4. Refactored the logic for resolving the qualified name first resolve by assignment (the common pattern)
5. `call_expr` is only used to access `func`, directly destructure it is a better approach. Thanks!
